File: GXX.CSharp/src/GXX.M2Server/Engine/UserCmds.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using GXX.M2Server.Engine;

namespace GXX.M2Server.Engine;

/// <summary>
/// UserCmds.pas TUserCmds 1:1（批次J57）：用户自定义命令表（UserCmd.txt，'命令名 TAB 编号'），
/// FCmdList 以 TStringList 承载 Objects[i]=编号。Find/Delete/Get 走 CompareText（不区分大小写），
/// GotoLable 跳转 '@UserCmd' + 编号（接缝）。
/// </summary>
public class TUserCmds
{
    private readonly List<string> _strings = new();
    private readonly List<int> _objects = new();
    private int _recordCount;

    public int Count => _strings.Count;

    public int RecordCount => _recordCount;

    public string GetStrings(int index) => _strings[index];

    public int GetObjects(int index) => _objects[index];

    public void SetObjects(int index, int value) => _objects[index] = value;

    /// <summary>LoadFromFile：UserCmd.txt，';' 行跳过，空名或编号 &lt; 0 丢弃。</summary>
    public void LoadFromFile()
    {
        _recordCount = 0;
        _strings.Clear();
        _objects.Clear();
        string sFileName = Path.Combine(M2Config.sEnvirDir, "UserCmd.txt");
        if (!File.Exists(sFileName))
            return;
        foreach (var line in File.ReadLines(sFileName, Encoding.GetEncoding(936)))
        {
            string sLineText = line;
            if (sLineText.Length == 0 || sLineText[0] == ';')
                continue;
            string sCmdName = "";
            string sIndex = "";
            sLineText = TSndaShopList.GetValidStr3(sLineText, ref sCmdName, ' ', '\t');
            sLineText = TSndaShopList.GetValidStr3(sLineText, ref sIndex, ' ', '\t');
            int nIndex = TSndaShopList.StrToIntDef(sIndex, -1);
            if (sCmdName.Length > 0 && nIndex >= 0)
            {
                _recordCount++;
                _strings.Add(sCmdName);
                _objects.Add(nIndex);
            }
        }
    }

    /// <summary>SaveToFile：每行 '命令名 TAB 编号'。</summary>
    public void SaveToFile()
    {
        var saveList = new List<string>();
        for (int i = 0; i < _strings.Count; i++)
            saveList.Add(_strings[i] + "\t" + _objects[i].ToString(System.Globalization.CultureInfo.InvariantCulture));
        string sFileName = Path.Combine(M2Config.sEnvirDir, "UserCmd.txt");
        try
        {
            File.WriteAllLines(sFileName, saveList, Encoding.GetEncoding(936));
        }
        catch
        {
            // Delphi try-finally + SaveToFile 异常向外（此处容错，保持数据不丢）
        }
    }

    public bool Add(string sCmd, int nIndex)
    {
        _strings.Add(sCmd);
        _objects.Add(nIndex);
        _recordCount++;
        return true;
    }

    public bool Delete(string sCmd)
    {
        for (int i = 0; i < _strings.Count; i++)
        {
            if (string.Compare(sCmd, _strings[i], StringComparison.OrdinalIgnoreCase) == 0)
            {
                _strings.RemoveAt(i);
                _objects.RemoveAt(i);
                _recordCount--;
                return true;
            }
        }
        return false;
    }

    public bool Delete(int nIndex)
    {
        for (int i = 0; i < _objects.Count; i++)
        {
            if (nIndex == _objects[i])
            {
                _strings.RemoveAt(i);
                _objects.RemoveAt(i);
                _recordCount--;
                return true;
            }
        }
        return false;
    }

    /// <summary>Get：CompareText 命中返回编号，否则 -1。</summary>
    public int Get(string sCmd)
    {
        for (int i = 0; i < _strings.Count; i++)
        {
            if (string.Compare(sCmd, _strings[i], StringComparison.OrdinalIgnoreCase) == 0)
                return _objects[i];
        }
        return -1;
    }

    public bool Find(string sCmd) => Get(sCmd) >= 0;

    public bool Find(int nIndex)
    {
        for (int i = 0; i < _objects.Count; i++)
        {
            if (nIndex == _objects[i])
                return true;
        }
        return false;
    }

    /// <summary>GotoLable 接缝（g_FunctionNPC.GotoLable(PlayObject, '@UserCmd'+Index, False) 等效）。</summary>
    public Action<object, string>? GotoLableHandler;

    /// <summary>g_FunctionNPC &lt;&gt; nil 门控（窗体/引擎层注入）。</summary>
    public Func<bool>? FunctionNPCExists;

    public void GotoLable(object playObject, int index)
    {
        if (index >= 0 && (FunctionNPCExists?.Invoke() ?? false))
            GotoLableHandler?.Invoke(playObject, "@UserCmd" + index.ToString(System.Globalization.CultureInfo.InvariantCulture));
    }

    public bool GotoLable(object playObject, string sCmd)
    {
        int index = Get(sCmd);
        if (index >= 0 && (FunctionNPCExists?.Invoke() ?? false))
        {
            GotoLableHandler?.Invoke(playObject, "@UserCmd" + index.ToString(System.Globalization.CultureInfo.InvariantCulture));
            return true;
        }
        return false;
    }

    /// <summary>窗体打开时使用的只读快照（RefUserCommandList 1:1）。</summary>
    public IEnumerable<(string Cmd, int Index)> Snapshot()
    {
        for (int i = 0; i < _strings.Count; i++)
            yield return (_strings[i], _objects[i]);
    }
}

File: Source/M2Engine/UserCmds.pas
unit UserCmds;

interface

uses
  Windows, Classes, SysUtils, ObjPlayer;

type
  TUserCmds = class
  private
    FRecordCount: Integer;
    FCmdList: TStringList;

    function GetCount: Integer;
    function GetObjects(Index: Integer): TObject;
    function GetStrings(Index: Integer): string;
  public
    constructor Create();
    destructor Destroy; override;

    procedure LoadFromFile;
    procedure SaveToFile;
    function Get(sCmd: string): Integer;
    function Find(sCmd: string): Boolean; overload;
    function Find(nIndex: Integer): Boolean; overload;
    function Add(sCmd: string; nIndex: Integer): Boolean;
    function Delete(sCmd: string): Boolean; overload;
    function Delete(nIndex: Integer): Boolean; overload;
    procedure GotoLable(PlayObject: TPlayObject; Index: Integer); overload;
    function GotoLable(PlayObject: TPlayObject; sCmd: string): Boolean; overload;

    property Strings[Index: Integer]: string read GetStrings;
    property Objects[Index: Integer]: TObject read GetObjects;
    property Count: Integer read GetCount;
    property RecordCount: Integer read FRecordCount;
  end;

implementation

uses
  M2Share, HUtil32;

constructor TUserCmds.Create();
begin
  FCmdList := TStringList.Create;
  FRecordCount := 0;
end;

destructor TUserCmds.Destroy;
begin
  FCmdList.Free;
  inherited;
end;

function TUserCmds.GetCount: Integer;
begin
  Result := FCmdList.Count;
end;

function TUserCmds.GetObjects(Index: Integer): TObject;
begin
  Result := FCmdList.Objects[Index];
end;

function TUserCmds.GetStrings(Index: Integer): string;
begin
  Result := FCmdList.Strings[Index];
end;

procedure TUserCmds.LoadFromFile;
var
  I: Integer;
  sFileName: string;
  sLineText: string;
  sCmdName: string;
  sIndex: string;
  nIndex: Integer;
  LoadList: TStringList;
begin
  FRecordCount := 0;
  FCmdList.Clear;
  sFileName := g_Config.sEnvirDir + 'UserCmd.txt';
  if FileExists(sFileName) then
  begin
    LoadList := TStringList.Create;
    try
      LoadList.LoadFromFile(sFileName);
      for I := 0 to LoadList.Count - 1 do
      begin
        sLineText := LoadList.Strings[I];
        if (sLineText <> '') and (sLineText[1] <> ';') then
        begin
          sLineText := GetValidStr3(sLineText, sCmdName, [' ', #9]);
          sLineText := GetValidStr3(sLineText, sIndex, [' ', #9]);
          nIndex := StrToIntDef(sIndex, -1);
          if (sCmdName <> '') and (nIndex >= 0) then
          begin
            Inc(FRecordCount);
            FCmdList.AddObject(sCmdName, TObject(nIndex));
          end;
        end;
      end;
    finally
      LoadList.Free;
    end;
  end;
end;

procedure TUserCmds.SaveToFile;
var
  I: Integer;
  sFileName: string;
  SaveList: TStringList;
begin
  SaveList := TStringList.Create;
  for I := 0 to FCmdList.Count - 1 do
  begin
    SaveList.Add(FCmdList.Strings[I] + #9 + IntToStr(Integer(FCmdList.Objects[I])));
  end;
  sFileName := g_Config.sEnvirDir + 'UserCmd.txt';
  try
    SaveList.SaveToFile(sFileName);
  finally
    SaveList.Free;
  end;

end;

function TUserCmds.Add(sCmd: string; nIndex: Integer): Boolean;
begin
  FCmdList.AddObject(sCmd, TObject(nIndex));
  Inc(FRecordCount);
  Result := True;
end;

function TUserCmds.Delete(sCmd: string): Boolean;
var
  I: Integer;
begin
  Result := False;
  for I := 0 to FCmdList.Count - 1 do
  begin
    if (CompareText(sCmd, FCmdList.Strings[I]) = 0) then
    begin
      FCmdList.Delete(I);
      Dec(FRecordCount);
      Result := True;
      Break;
    end;
  end;
end;

function TUserCmds.Delete(nIndex: Integer): Boolean;
var
  I: Integer;
begin
  Result := False;
  for I := 0 to FCmdList.Count - 1 do
  begin
    if nIndex = Integer(FCmdList.Objects[I]) then
    begin
      FCmdList.Delete(I);
      Dec(FRecordCount);
      Result := True;
      Break;
    end;
  end;
end;

function TUserCmds.Get(sCmd: string): Integer;
var
  I: Integer;
begin
  Result := -1;
  for I := 0 to FCmdList.Count - 1 do
  begin
    if CompareText(sCmd, FCmdList.Strings[I]) = 0 then
    begin
      Result := Integer(FCmdList.Objects[I]);
      Break;
    end;
  end;
end;

function TUserCmds.Find(sCmd: string): Boolean;
var
  I: Integer;
begin
  Result := False;
  for I := 0 to FCmdList.Count - 1 do
  begin
    if CompareText(sCmd, FCmdList.Strings[I]) = 0 then
    begin
      Result := True;
      Break;
    end;
  end;
end;

function TUserCmds.Find(nIndex: Integer): Boolean;
var
  I: Integer;
begin
  Result := False;
  for I := 0 to FCmdList.Count - 1 do
  begin
    if nIndex = Integer(FCmdList.Objects[I]) then
    begin
      Result := True;
      Break;
    end;
  end;
end;

procedure TUserCmds.GotoLable(PlayObject: TPlayObject; Index: Integer);
begin
  if (Index >= 0) and (g_FunctionNPC <> nil) then
  begin
    PlayObject.m_nScriptGotoCount := 0;
    g_FunctionNPC.GotoLable(PlayObject, '@UserCmd' + IntToStr(Index), False);
  end;
end;

function TUserCmds.GotoLable(PlayObject: TPlayObject; sCmd: string): Boolean;
var
  Index: Integer;
begin
  Result := False;
  Index := Get(sCmd);
  if (Index >= 0) and (g_FunctionNPC <> nil) then
  begin
    PlayObject.m_nScriptGotoCount := 0;
    g_FunctionNPC.GotoLable(PlayObject, '@UserCmd' + IntToStr(Index), False);
    Result := True;
  end;
end;

end.
